cypress/plugins/nodeManager/plugin.ts

```diff
@@ -0,0 +1,345 @@
+import * as path from "path";
+import * as childProcess from "child_process";
+import fetch from "node-fetch";
+import waitOn from "wait-on";
+import type { NodeSession } from "./shared";
+import { Commands } from "./shared";
+
+type NodeId = number;
+type PeerAddress = string;
+type AuthToken = string;
+
+const ROOT_PATH = path.join(__dirname, "../../../");
+
+// IP to which all started processes will bind to.
+const HOST = "127.0.0.1";
+
+const sleep = async (ms: number) => {
+  await new Promise(resolve => setTimeout(resolve, ms));
+};
+
+class Logger {
+  prefix: string;
+  indentationLevel: number;
+
+  constructor({ prefix = "", indentationLevel = 1 }) {
+    this.prefix = prefix;
+    this.indentationLevel = indentationLevel;
+  }
+
+  log(message: string) {
+    const indentation = " ".repeat(this.indentationLevel * 2);
+    console.log(indentation + this.prefix + message);
+  }
+}
+
+// Because it's not possible to mix tagged union types and extending
+// interfaces, we have to repeat the "inherited" attributes in each node state.
+interface ConfiguredNode {
+  kind: "configured";
+}
+
+interface StartedNode {
+  kind: "started";
+  process: childProcess.ChildProcess;
+}
+
+interface OnboardedNode {
+  kind: "onboarded";
+  process: childProcess.ChildProcess;
+  authToken: AuthToken;
+  peerAddress: PeerAddress;
+}
+
+type NodeState = ConfiguredNode | StartedNode | OnboardedNode;
+
+class Node {
+  private state: NodeState = { kind: "configured" };
+  private logger: Logger;
+
+  id: NodeId;
+  httpPort: number;
+  peerPort: number;
+  proxyBinaryPath: string;
+
+  get peerAddress(): PeerAddress {
+    if (this.state.kind !== "onboarded") {
+      throw new Error("Can't get peerAddress before node is onboarded");
+    }
+
+    return this.state.peerAddress;
+  }
+
+  get authToken(): AuthToken {
+    if (this.state.kind !== "onboarded") {
+      throw new Error("Can't get peerAddress before node is onboarded");
```

Comment:
This OOP state stuff feels off to me. I'd expect a cleaner solution out of having state transition functions:
```
configure : ConfiguredNode
start : ConfiguredNode -> Result<StartedNode>
onboard: StartedNode -> Result<OnboardedNode>
```

then your `OnboardedNode` functions would _only_ work on those. I dunno how easy this is to represent in TS but I feel like it's possible. I imagine the `Node` would be generic over `T` where that would be filled in by our different node states. This rules out these state-specific errors.

Food for thought and maybe worth a follow-up ticket.

Reply:
> This OOP state stuff feels off to me.

I agree, maybe we can remove the inner Node class, but then the NodeManager would have to keep an array for each different state.

Will do a follow up once I know how this can be done better in Typescript.

Reply:
Ya, that's a good point. But could be interesting in that it only needs to inspect the specific arrays when want to promote nodes :)

> Will do a follow up once I know how this can be done better in Typescript.
:ok_hand: